ui/app/src/state/reducers/pathNavigatorTree.ts

```diff
@@ -0,0 +1,162 @@
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+import { createReducer } from '@reduxjs/toolkit';
+import { PathNavigatorTreeStateProps } from '../../components/PathNavigatorTree';
+import LookupTable from '../../models/LookupTable';
+import {
+  pathNavigatorTreeCollapsePath,
+  pathNavigatorTreeExpandPath,
+  pathNavigatorTreeFetchPathChildren,
+  pathNavigatorTreeFetchPathChildrenComplete,
+  pathNavigatorTreeFetchPathPageComplete,
+  pathNavigatorTreeInit,
+  pathNavigatorTreeRestoreComplete,
+  pathNavigatorTreeSetKeyword,
+  pathNavigatorTreeToggleExpanded
+} from '../actions/pathNavigatorTree';
+import { changeSite } from './sites';
+
+const reducer = createReducer<LookupTable<PathNavigatorTreeStateProps>>(
+  {},
+  {
+    [pathNavigatorTreeInit.type]: (state, { payload: { id, path, collapsed = false, limit, expanded = [] } }) => {
+      return {
+        ...state,
+        [id]: {
+          rootPath: path,
+          levelDescriptor: null,
+          collapsed,
+          limit,
+          expanded,
+          childrenByParentPath: {},
+          keywordByPath: {},
+          totalByPath: {}
+        }
+      };
+    },
+    [pathNavigatorTreeExpandPath.type]: (state, { payload: { id, path } }) => {
+      return {
+        ...state,
+        [id]: {
+          ...state[id],
+          expanded: [...state[id].expanded, path]
+        }
+      };
+    },
+    [pathNavigatorTreeCollapsePath.type]: (state, { payload: { id, path } }) => {
+      return {
+        ...state,
+        [id]: {
+          ...state[id],
+          expanded: state[id].expanded.filter((expanded) => expanded !== path)
+        }
+      };
+    },
+    [pathNavigatorTreeToggleExpanded.type]: (state, { payload: { id, collapsed } }) => {
+      return {
+        ...state,
+        [id]: {
+          ...state[id],
+          collapsed
+        }
+      };
+    },
+    [pathNavigatorTreeSetKeyword.type]: (state, { payload: { id, path, keyword } }) => {
+      return {
+        ...state,
+        [id]: {
+          ...state[id],
+          keywordByPath: {
+            ...state[id].keywordByPath,
+            [path]: keyword
+          }
+        }
+      };
+    },
+    [pathNavigatorTreeFetchPathChildren.type]: (state, { payload: { id, path } }) => {
+      return {
+        ...state,
+        [id]: {
+          ...state[id],
+          expanded: state[id].expanded.includes(path) ? [...state[id].expanded] : [...state[id].expanded, path]
+        }
+      };
+    },
+    [pathNavigatorTreeFetchPathChildrenComplete.type]: (state, { payload: { id, parentPath, children, options } }) => {
+      return {
+        ...state,
+        [id]: {
+          ...state[id],
+          ...(children.length === 0 &&
+            !options?.keyword && { expanded: state[id].expanded.filter((path) => path !== parentPath) }),
```

Comment:
```suggestion
          expanded: children.length === 0 && !options?.keyword ? state[id].expanded.filter((path) => path !== parentPath) }) : state[id].expanded,
```

What's this logic for though? Why collapse the path when there aren't children or keyword?

Reply:
I still think the ternary in the suggestion is more efficient than a spread of a single prop.